### PR TITLE
Add gradient plasticity to the crystal models

### PR DESCRIPTION
**Pull request by reverendbedford**
This pull will add gradient plasticity (ala the Nye tensor) to the crystal models.  Of course we can't actually calculate the Nye tensor internal to NEML, but it provides the necessary information (inverse elastic tensor) and now has a mechanism to accept the calculated Nye tensor from some higher level "thing" (i.e. MOOSE/FEA).

The PR also updates some of the slip hardening models to have Nye contributions.  These default to zero so that the default behavior is the same as our reference models.

I should have made a separate PR, but this also includes an update to our polycrystal homogenization moving it to the C++ side and inheriting from the NEMLModel_ld interface so that polcrystal models are now just regular NEML models.  This means they can be driven/fit from the python interface and loaded and used in MOOSE.

### CODE REVIEW

File: doc/sphinx/crystal_plasticity.rst [hunk 1]
```diff
@@ -205,6 +205,22 @@ orientation update equations must be supplemented with the evolution
...
+
+Describing the areal density of geometrically necessary dislocations.
+The Nye tensor can be used as an *ad hoc* contributor to slip system
+flow or hardening models.  The NEML system itself cannot calculate
+the Nye tensor itself -- that requires gradient information not available
+at the material point level -- but can accept the calculated Nye tensor
+from a higher level framework, for example an FEA model, and process it
+accordingly.
+To help with this calculation, the model provides the inverse elastic
+tensor :math:`\mathbf{F}^{e-1}` as an available output.
```
Comment (chakra34) [inline doc/sphinx/crystal_plasticity.rst]: We need to mention in which configuration is the curl taken, I believe in this definition it is in the current configuration, for reference one can look into Das, Hofmann and Tarleton 2018.

Comment (reverendbedford) [reply doc/sphinx/crystal_plasticity.rst]: Yes, it's the current configuration.  I'll note it in the docs.

File: doc/sphinx/crystal_plasticity.rst [hunk 2]
```diff
@@ -205,6 +205,22 @@ orientation update equations must be supplemented with the evolution
...
 
+In addition to this basic kinematic framework, NEML has the ability
+to provide crystal models with the Nye tensor
+
+.. math::
+   \bm{\alpha} = -\nabla \times \mathbf{F}^{e-1}
+
+Describing the areal density of geometrically necessary dislocations.
+The Nye tensor can be used as an *ad hoc* contributor to slip system
+flow or hardening models.  The NEML system itself cannot calculate
```
[the remark this answers is not in the document]
Comment (reverendbedford) [reply doc/sphinx/crystal_plasticity.rst]: Yep

File: doc/sphinx/cp/polycrystal/taylor.rst [hunk 3]
```diff
@@ -0,0 +1,39 @@
+TaylorModel
+===========
+
+Overview
+--------
+
+This polycrystal homogenization model implements the standard Taylor approximation.  Here the individual crystal receiver the same deformation information and the resulting stresses are averaged.
+
```
[the remark this answers is not in the document]
Comment (reverendbedford) [reply doc/sphinx/cp/polycrystal/taylor.rst]: Yep

File: doc/sphinx/cp/SlipRule.rst [hunk 4]
```diff
@@ -8,9 +8,9 @@ These objects provide a relation between the stress, history, and temperature
 and the slip rate on each individual slip system as well as the history evolution.  The interface used is:
 
 .. math::
-   \dot{\gamma}_{g,i}, \frac{\partial \dot{\gamma}_{g,i}}{\partial \bm{\sigma}}, \frac{\partial \dot{\gamma}_{g,i}}{\partial \mathbf{h}} \leftarrow \mathcal{G}\left( \bm{\sigma}, \mathbf{h}, T \right)
+   \dot{\gamma}_{g,i}, \frac{\partial \dot{\gamma}_{g,i}}{\partial \bm{\sigma}}, \frac{\partial \dot{\gamma}_{g,i}}{\partial \mathbf{h}} \leftarrow \mathcal{G}\left( \bm{\sigma}, \mathbf{h}, \bm{\alpha}, T \right)
 
-   \dot{\mathbf{h}}, \frac{\partial \dot{\mathbf{h}}}{\partial \bm{\sigma}}, \frac{\partial \dot{\mathbf{h}}}{\partial \mathbf{h}} \leftarrow \mathcal{H}\left(\bm{\sigma}, \mathbf{h}, T \right)
+   \dot{\mathbf{h}}, \frac{\partial \dot{\mathbf{h}}}{\partial \bm{\sigma}}, \frac{\partial \dot{\mathbf{h}}}{\partial \mathbf{h}} \leftarrow \mathcal{H}\left(\bm{\sigma}, \mathbf{h}, \bm{\alpha}, T \right)
 
 where :math:`g` indicates the slip family and :math:`i` indicates the system within the group.
```
[the remark this answers is not in the document]
Comment (reverendbedford) [reply doc/sphinx/cp/SlipRule.rst]: Yep

File: doc/sphinx/cp/sliprule/SlipStrengthSlipRule.rst [hunk 5]
```diff
@@ -13,9 +13,9 @@ and the slip rate on each individual slip system where the slip rate is related
...
 
 .. math::
-   \dot{\gamma}_{g,i}, \frac{\partial \dot{\gamma}_{g,i}}{\partial \tau_{g,i}}, \frac{\partial \dot{\gamma}_{g,i}}{\partial \bar{\tau}_{g,i}} \leftarrow \mathcal{G}\left( \tau_{g,i}, \bar{\tau}_{g,i}, T \right)
+   \dot{\gamma}_{g,i}, \frac{\partial \dot{\gamma}_{g,i}}{\partial \tau_{g,i}}, \frac{\partial \dot{\gamma}_{g,i}}{\partial \bar{\tau}_{g,i}} \leftarrow \mathcal{G}\left( \tau_{g,i}, \bar{\tau}_{g,i}, \bm{\alpha}, T \right)
 
-   \dot{\mathbf{h}}, \frac{\partial \dot{\mathbf{h}}}{\partial \bm{\sigma}}, \frac{\partial \dot{\mathbf{h}}}{\partial \mathbf{h}} \leftarrow \mathcal{H}\left(\bm{\sigma}, \mathbf{h}, T \right)
+   \dot{\mathbf{h}}, \frac{\partial \dot{\mathbf{h}}}{\partial \bm{\sigma}}, \frac{\partial \dot{\mathbf{h}}}{\partial \mathbf{h}} \leftarrow \mathcal{H}\left(\bm{\sigma}, \mathbf{h}, \bm{\alpha}, T \right)
 
 where :math:`g` indicates the slip group, :math:`i` indicates the system within the group, and :math:`\bar{\tau}_{g,i}` is the slip system strength, defined by a SlipHardening model:
```
[the remark this answers is not in the document]
Comment (reverendbedford) [reply doc/sphinx/cp/sliprule/SlipStrengthSlipRule.rst]: We'll stick with group.

File: examples/cp/flow_curve.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    pmodel = polycrystal.TaylorModel(model, orientations, nthreads = nthreads)
+
+    res = drivers.uniaxial_test(pmodel, erate, emax = emax, nsteps = steps, 
+        verbose = True)
+
+    strains.append(res['strain'])
+    stresses.append(res['stress'])
+  
+  plt.style.use("presentation")
```
[the remark this answers is not in the document]
Comment (reverendbedford) [reply examples/cp/flow_curve.py]: Yes, that is my fault for copy/pasting from another script.

**Pull request merged**